sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidRexExecutor.java

```diff
@@ -142,40 +142,42 @@ public void reduce(
-                                                                final Number longVal = (Number) val;
-                                                                if (longVal == null) {
-                                                                  return null;
-                                                                }
-                                                                return BigDecimal.valueOf(longVal.longValue());
-                                                              })
-                                                              .collect(Collectors.toList());
+              List<BigDecimal> resultAsBigDecimalList = new ArrayList<>(array.length);
+              for (Object val : array) {
+                final Number longVal = (Number) val;
+                if (longVal == null) {
+                  resultAsBigDecimalList.add(null);
+                } else {
+                  resultAsBigDecimalList.add(BigDecimal.valueOf(longVal.longValue()));
+                }
+              }
               literal = rexBuilder.makeLiteral(resultAsBigDecimalList, constExp.getType(), true);
             } else {
-              List<BigDecimal> resultAsBigDecimalList = Arrays.stream(exprResult.asArray()).map(
-                  val -> {
-                    final Number doubleVal = (Number) val;
-                    if (doubleVal == null) {
-                      return null;
-                    }
-                    if (Double.isNaN(doubleVal.doubleValue()) || Double.isInfinite(doubleVal.doubleValue())) {
-                      String expression = druidExpression.getExpression();
-                      throw new UnsupportedSQLQueryException(
-                          "'%s' contains an element that evaluates to '%s' which is not supported in SQL. You can either cast the element in the array to bigint or char or change the expression itself",
-                          expression,
-                          Double.toString(doubleVal.doubleValue())
-                      );
-                    }
-                    return BigDecimal.valueOf(doubleVal.doubleValue());
-                  }
-              ).collect(Collectors.toList());
+              List<BigDecimal> resultAsBigDecimalList = new ArrayList<>(array.length);
+              for (Object val : array) {
+                final Number doubleVal = (Number) val;
+                if (doubleVal == null) {
+                  resultAsBigDecimalList.add(null);
+                } else if (Double.isNaN(doubleVal.doubleValue()) || Double.isInfinite(doubleVal.doubleValue())) {
+                  String expression = druidExpression.getExpression();
+                  throw new UnsupportedSQLQueryException(
+                      "'%s' contains an element that evaluates to '%s' which is not supported in SQL. You can either cast the element in the array to bigint or char or change the expression itself",
```

Comment:
I know you didn't change anything here, but you moved it so I get to comment 🙂

"BIGINT or VARCHAR" would be better than "bigint or char".